Google Analytics for B&C Search

mmcgowanBOS opened the pull request:
- Added Google Analytics node module.
- Added front end js to REACT app.

mmcgowanBOS committed:
- Added Google Analytics node module.
- Added front end js to REACT app

fionawhim commented on services-js/commissions-search/src/App.js:
```diff
@@ -1,10 +1,12 @@
 import ResultList from './ResultList';
+import ReactGA from 'react-ga';
+
+ReactGA.initialize('UA-2187282-27');
```
It’s not good to hard-code values like this into source code.

In theory configuration values like this should be made available by something in our Drupal template. See how we're doing that for the API key here: https://github.com/CityOfBoston/digital/blob/develop/services-js/commissions-search/src/index.js#L25

Though since you seem to want to use the same GA property as Boston.gov already has, do you really need `ReactGA`, as opposed to just assuming that the Google Analytics stuff is already loaded?

I'm worried that this will lead to double-counting.

mmcgowanBOS replied:
Removed GA module. Now using dataLayer Object to push data to GA. 
*Note, this approach does not work with existing Google Tag Manager (GTM) included on the Drupal (or any other) template page UNLESS specific Tags, Triggers, and Variables are set up to integrate properly in GTM.

fionawhim commented on services-js/commissions-search/src/App.js:
```diff
@@ -50,13 +52,40 @@ class App extends React.Component {
   };
 
+  findCheckedBoxes4GA = event => {
+    var chk_arr = document.getElementsByClassName('cb');
```
A few style things:

 - Use camelCase for JavaScript identifiers.
 - Don’t use `var`. Use `const` by default, `let` if you need to.

mmcgowanBOS replied:
sure.

fionawhim commented on services-js/commissions-search/src/App.js:
```diff
@@ -50,13 +52,40 @@ class App extends React.Component {
+    var chk_arr = document.getElementsByClassName('cb');
+    var chklength = chk_arr.length;
+    for (var k = 0; k < chklength; k++) {
+      var checkit = chk_arr[k].firstChild.checked ? true : false;
```
This data should already be present in the component’s state, so you should get it from there rather than inspecting HTML elements.

Also, avoid `document`-based selection in general in React, since you could be grabbing input elements from other places on the page. If you reeeeeaaaaalllly need native DOM operations, use a Ref to get access to an HTML element specifically owned by your component.

mmcgowanBOS replied:
Per our meeting / conversation, this was / is not available in the component state and would have to be altered using a rather buried Apollo set up.

fionawhim replied:
Can you put that in the code as a comment?

It's much more likely to be found there when someone needs to look back on the code than digging up the old PR.

mmcgowanBOS replied:
I ended up adding an attribute to the checkbox input elements in the FacetList.js file that displays the label value. Having these values available directly in the DOM we now can easily access them and add to REACT state in the App.js file. From here we can iterate over the new state Object (checkedNames) upon submission of the filter and send to GA.

fionawhim commented on services-js/commissions-search/src/App.js:
```diff
@@ -50,13 +52,40 @@ class App extends React.Component {
+    }
+  };
+
+  sendEvent2GA = (actionV, categoryV, labelV) => {
```
good encapsulation.

mmcgowanBOS committed:
- Removed references to GA node module.
- Updated method to send / push data to GA using dataLayer.push
- Included GTM code for index.html files so local testing of application will still work

fionawhim commented on services-js/commissions-search/src/App.js:
```diff
@@ -53,12 +50,12 @@ class App extends React.Component {
-      var checkit = chk_arr[k].firstChild.checked ? true : false;
-      if (checkit === true) {
-        var children = chk_arr[k].childNodes;
+    const chkArr = document.getElementsByClassName('cb');
```
I know we talked about commenting this code (to explain why you’re not using state) and using something lest fragile than `firstChild` and `children[1]`. 

Is there a reason you didn't go that route?

mmcgowanBOS replied:
I was waiting for Travis to finish and then going to make comments in GitHub and address the questions / comments there.

mmcgowanBOS replied:
See comments above for updates to this.

mmcgowanBOS committed:
- Added attribute for input boxes to display label value
- Updated App to parse out label value and add to State

fionawhim commented on services-js/commissions-search/package.json:
```diff
@@ -12,6 +12,7 @@
     "react": "16.6.0",
     "react-apollo": "^2.1.9",
     "react-dom": "16.6.0",
+    "react-ga": "^2.5.6",
```
Remove this, since it's not being used, and be sure to re-run `yarn install` so it gets taken out of `yarn.lock`

fionawhim commented on services-js/commissions-search/src/App.js:
```diff
@@ -30,33 +31,63 @@ class App extends React.Component {
 
+  //get checkbox names from state and send to GA
+  findCheckedBoxes4GA = event => {
+    const chkArr = { ...this.state.checkedNames };
```
You don't need to splat this into a new Object.

Also, `Arr` is not a good name for this since it's an object, not an array.

fionawhim commented on services-js/commissions-search/src/FacetList.js:
```diff
@@ -25,6 +25,7 @@ function renderAreaCheckbox(props, policyArea) {
     <Checkbox
       key={`checkbox-area-${policyArea.id}`}
       name={policyArea.id}
+      nameval={labelText}
```
I'd recommend using `data-` for custom attributes. (E.g. `data-policy-name={labelText}`)

fionawhim commented:
Approved per Kayla’s request.

mmcgowanBOS committed:
Removed reference to react-ga in package file and yarn.lock
